File: src/1420.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MODULO (int)(1e9 + 7)
int numOfArrays(int n, int m, int k) {
    int retVal = 0;

    long long dp[m + 1][k + 1];
    memset(dp, 0, sizeof(dp));
    long long prefix[m + 1][k + 1];
    memset(prefix, 0, sizeof(prefix));
    long long prevDp[m + 1][k + 1];
    memset(prevDp, 0, sizeof(prevDp));
    long long prevPrefix[m + 1][k + 1];
    memset(prevPrefix, 0, sizeof(prevPrefix));

    int i;
    for (i = 1; i <= m; i++) {
        dp[i][1] = 1;
    }
    long long answer;
    int maxNum, cost;
    for (i = 1; i <= n; i++) {
        if (i > 1) {
            memset(dp, 0, sizeof(dp));
        }

        memset(prefix, 0, sizeof(prefix));
        for (maxNum = 1; maxNum <= m; maxNum++) {
            for (cost = 1; cost <= k; cost++) {
                answer = (maxNum * prevDp[maxNum][cost]) % MODULO;
                answer = (answer + prevPrefix[maxNum - 1][cost - 1]) % MODULO;

                dp[maxNum][cost] += answer;
                dp[maxNum][cost] %= MODULO;

                prefix[maxNum][cost] = (prefix[maxNum - 1][cost] + dp[maxNum][cost]);
                prefix[maxNum][cost] %= MODULO;
            }
        }

        for (maxNum = 0; maxNum <= m; maxNum++) {
            for (cost = 0; cost <= k; cost++) {
                prevDp[maxNum][cost] = dp[maxNum][cost];
                prevPrefix[maxNum][cost] = prefix[maxNum][cost];
            }
        }
    }
    retVal = prefix[m][k];

    return retVal;
}

int main(int argc, char** argv) {
    struct testCaseType {
        int n;
        int m;
        int k;
    } testCase[] = {{2, 3, 1}, {5, 2, 3}, {9, 1, 1}};
    int numberOfTestCase = sizeof(testCase) / sizeof(testCase[0]);
    /* Example
     *  Input: n = 2, m = 3, k = 1
     *  Output: 6
     *
     *  Input: n = 5, m = 2, k = 3
     *  Output: 0
     *
     *  Input: n = 9, m = 1, k = 1
     *  Output: 1
     */

    int answer;
    int i;
    for (i = 0; i < numberOfTestCase; ++i) {
        printf("Input: n = %d, m = %d, k = %d\n", testCase[i].n, testCase[i].m, testCase[i].k);

        answer = numOfArrays(testCase[i].n, testCase[i].m, testCase[i].k);
        printf("Output: %d\n", answer);

        printf("\n");
    }

    return EXIT_SUCCESS;
}

File: src/1420.py
import sys
import logging
from typing import List, Optional
from collections import defaultdict, deque, Counter
from functools import cache, lru_cache
from bisect import bisect_left, bisect_right
from itertools import accumulate
from heapq import heapify, heappush, heappop


def logging_setting():
    LOG_LEVEL = logging.INFO    # DEBUG, INFO, WARNING, ERROR, CRITICAL
    if __debug__:
        LOG_LEVEL = logging.DEBUG

    logging.basicConfig(
        stream=sys.stderr,
        level=LOG_LEVEL,
        format="%(levelname)-6s %(asctime)s %(module)s:%(funcName)s:%(lineno)-4s %(message)s",
        datefmt='%Y/%m/%d %H:%M:%S')


class Solution:
    def __init__(self):
        self.MODULO = 10 ** 9 + 7

    def numOfArrays(self, n: int, m: int, k: int) -> int:
        retVal = 0

        dp = [[0] * (k + 1) for _ in range(m + 1)]
        prefix = [[0] * (k + 1) for _ in range(m + 1)]
        prevDp = [[0] * (k + 1) for _ in range(m + 1)]
        prevPrefix = [[0] * (k + 1) for _ in range(m + 1)]

        for num in range(1, m + 1):
            dp[num][1] = 1
        for i in range(1, n + 1):
            if i > 1:
                dp = [[0] * (k + 1) for _ in range(m + 1)]

            prefix = [[0] * (k + 1) for _ in range(m + 1)]
            for max_num in range(1, m + 1):
                for cost in range(1, k + 1):
                    answer = (max_num * prevDp[max_num][cost]) % self.MODULO
                    answer = (answer + prevPrefix[max_num - 1][cost - 1]) % self.MODULO

                    dp[max_num][cost] += answer
                    dp[max_num][cost] %= self.MODULO

                    prefix[max_num][cost] = (prefix[max_num - 1][cost] + dp[max_num][cost]) % self.MODULO

            prevDp = dp
            prevPrefix = prefix

        retVal = prefix[m][k]

        return retVal


if __name__ == "__main__":
    logging_setting()

    try:
        logging.info("sys.version: %s", sys.version)
        print()

        pSolution = Solution()
        for n, m, k in zip([2, 5, 9], [3, 2, 1], [1, 3, 1]):
            # /* Example
            #  *  Input: n = 2, m = 3, k = 1
            #  *  Output: 6
            #  *
            #  *  Input: n = 5, m = 2, k = 3
            #  *  Output: 0
            #  *
            #  *  Input: n = 9, m = 1, k = 1
            #  *  Output: 1
            #  */
            logging.info("Input: n = %s, m = %s, k = %s", n, m, k)

            retVal = pSolution.numOfArrays(n, m, k)
            logging.info("Output: %s", retVal)

            print()
    except KeyboardInterrupt as exception:
        logging.error("%s: %s", exception.__class__.__name__, exception, exc_info=True)
        pass
    except Exception as exception:
        logging.error("%s: %s", exception.__class__.__name__, exception, exc_info=True)
        pass

    sys.exit(0)
